support/refc/memoryManagement.c

```diff
@@ -134,39 +188,45 @@ Value_Array *idris2_makeArray(int length) {
 }
 
 Value *idris2_newReference(Value *source) {
+  IDRIS2_INC_MEMSTAT(n_newReference);
   // note that we explicitly allow NULL as source (for erased arguments)
-  if (source && !idris2_vp_is_unboxed(source)) {
-    source->header.refCounter++;
+  if (source && !idris2_vp_is_unboxed(source) &&
+      source->header.refCounter != IDRIS2_VP_REFCOUNTER_MAX) {
+    IDRIS2_INC_MEMSTAT(n_actualNewReference);
+    ++source->header.refCounter;
+    if (source->header.refCounter == IDRIS2_VP_REFCOUNTER_MAX)
+      IDRIS2_INC_MEMSTAT(n_immortalized);
   }
   return source;
 }
 
 void idris2_removeReference(Value *elem) {
-  if (!elem || idris2_vp_is_unboxed(elem)) {
+  IDRIS2_INC_MEMSTAT(n_removeReference);
+  if (!elem || idris2_vp_is_unboxed(elem))
     return;
-  }
-
-  IDRIS2_REFC_VERIFY(elem->header.refCounter > 0, "refCounter %lld",
-                     (long long)elem->header.refCounter);
-  // remove reference counter
-  elem->header.refCounter--;
-  if (elem->header.refCounter == 0)
-  // recursively remove all references to all children
-  {
+  else if (elem->header.refCounter == IDRIS2_VP_REFCOUNTER_MAX) {
+    IDRIS2_INC_MEMSTAT(n_tried_to_kill_immortals);
+    return;
+  } else if (elem->header.refCounter != 1) {
+    --elem->header.refCounter;
+    return;
+  } else {
+    IDRIS2_INC_MEMSTAT(n_freed);
```

Comment:
I notice that whereas `elem->header.refCounter` used to be decremented prior to freeing, now it is not; that seems theoretically ok to me, but I wanted to make sure it was intentional because I have not verified that nothing downstream of this code would double check the ref count.

Reply:
I decided not to write back the 0 value of the counter because I thought the overhead of updating the contents of a deleted object would be wasted.

The improvements are so small that they have little real effect.